src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -697,67 +695,75 @@ private static int WriteLeftAlignedBytesAndGetChecksum(ReadOnlySpan<byte> bytesT
+            bytesToWrite.CopyTo(destination);
 
-            return checksum;
+            return Checksum(bytesToWrite);
         }
 
         // Writes the specified bytes aligned to the right, filling all the leading bytes with the zero char 0x30,
         // ensuring a null terminator is included at the end of the specified span.
         private static int WriteRightAlignedBytesAndGetChecksum(ReadOnlySpan<byte> bytesToWrite, Span<byte> destination)
         {
-            int checksum = 0;
-            int i = destination.Length - 1;
-            int j = bytesToWrite.Length - 1;
+            Debug.Assert(destination.Length > 1);
 
-            while (i >= 0)
+            // Null terminated
+            destination[^1] = (byte)'\0';
+
+            // Copy as many input bytes as will fit
+            int numToCopy = Math.Min(bytesToWrite.Length, destination.Length - 1);
+            bytesToWrite = bytesToWrite.Slice(0, numToCopy);
+            int copyPos = destination.Length - 1 - bytesToWrite.Length;
+            bytesToWrite.CopyTo(destination.Slice(copyPos));
+
+            // Fill all leading bytes with zeros
+            destination.Slice(0, copyPos).Fill((byte)'0');
+
+            return Checksum(destination);
+        }
+
+        private static int Checksum(ReadOnlySpan<byte> bytes)
+        {
+            int checksum = 0;
+            foreach (byte b in bytes)
             {
-                if (i == destination.Length - 1)
-                {
-                    destination[i] = 0; // null terminated
-                }
-                else if (j >= 0)
-                {
-                    destination[i] = bytesToWrite[j];
-                    j--;
-                }
-                else
-                {
-                    destination[i] = TarHelpers.ZeroChar; // leading zeros
-                }
-                checksum += destination[i];
-                i--;
+                checksum += b;
             }
-
             return checksum;
         }
 
         // Writes the specified decimal number as a right-aligned octal number and returns its checksum.
-        internal static int WriteAsOctal(long tenBaseNumber, Span<byte> destination, int location, int length)
+        internal static int FormatOctal(long value, Span<byte> destination)
         {
-            long octal = TarHelpers.ConvertDecimalToOctal(tenBaseNumber);
-            byte[] bytes = Encoding.ASCII.GetBytes(octal.ToString());
-            return WriteRightAlignedBytesAndGetChecksum(bytes.AsSpan(), destination.Slice(location, length));
+            ulong remaining = (ulong)value;
+            Span<byte> digits = stackalloc byte[32]; // longer than any possible octal formatting of a ulong
+
+            int i = digits.Length - 1;
+            while (true)
+            {
+                digits[i] = (byte)('0' + (remaining % 8));
+                remaining /= 8;
+                if (remaining == 0) break;
+                i--;
+            }
```

Comment:
SharpLab show that there are no optimizations with shifts as I'd expect. So why not unsafe convert the long to span<byte>[8]?